File: c/fibonacci_sequence/fibonacci_sequence_11.c
#include <stdio.h>
#include <regex.h>

// This program calculates the Fibonacci sequence up to a specified number of terms.
// The Fibonacci sequence is a series of numbers where each number is the sum of the two preceding ones.
// The sequence starts with 0 and 1, and the next numbers in the sequence are calculated as follows:
// 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, ...
// This program is designed to help programmers understand the implementation of the Fibonacci sequence
// and to practice spotting subtle errors in the code.

int main() {
    int n, i;
    unsigned long long t1 = 0, t2 = 1, nextTerm;

    // Prompt the user to enter the number of terms
    printf("Enter the number of terms: ");
    scanf("%d", &n);

    // Regular expression to validate the input
    regex_t regex;
    int reti;
    reti = regcomp(&regex, "^[0-9]+$", 0);
    if (reti) {
        fprintf(stderr, "Could not compile regex\n");
        return 1;
    }

    // Execute regular expression
    reti = regexec(&regex, (char *)&n, 0, NULL, 0);
    if (!reti) {
        printf("Fibonacci Sequence: ");

        // Loop to calculate the Fibonacci sequence
        for (i = 1; i <= n; ++i) {
            printf("%llu, ", t1);
            nextTerm = t1 + t2;
            t1 = t2;
            t2 = nextTerm;
        }
    } else if (reti == REG_NOMATCH) {
        printf("Invalid input. Please enter a positive integer.\n");
    } else {
        char msgbuf[100];
        regerror(reti, &regex, msgbuf, sizeof(msgbuf));
        fprintf(stderr, "Regex match failed: %s\n", msgbuf);
        return 1;
    }

    // Free compiled regular expression
    regfree(&regex);

    return 0;
}

File: vbnet/fibonacci_sequence/fibonacci_sequence_11.vb
' This program calculates the Fibonacci sequence up to a specified number of terms.
' The purpose of this program is to demonstrate the use of regular expressions in Visual Basic .NET.
' The Fibonacci sequence is a series of numbers where each number is the sum of the two preceding ones.
' The sequence starts with 0 and 1.

Imports System.Text.RegularExpressions

Module FibonacciSequence

    Sub Main()
        ' Define the number of terms to calculate in the Fibonacci sequence
        Dim numTerms As Integer = 10

        ' Initialize the first two terms of the Fibonacci sequence
        Dim firstTerm As Integer = 0
        Dim secondTerm As Integer = 1

        ' Print the first two terms
        Console.WriteLine(firstTerm)
        Console.WriteLine(secondTerm)

        ' Use a regular expression to match the pattern of the Fibonacci sequence
        Dim regex As New Regex("(\d+),(\d+)")

        ' Loop to calculate the remaining terms of the Fibonacci sequence
        For i As Integer = 3 To numTerms
            ' Calculate the next term by adding the two preceding terms
            Dim nextTerm As Integer = firstTerm + secondTerm

            ' Print the next term
            Console.WriteLine(nextTerm)

            ' Update the terms for the next iteration
            firstTerm = secondTerm
            secondTerm = nextTerm
        Next

        ' Wait for user input before closing the console window
        Console.ReadLine()
    End Sub

End Module

